File: langtonsAnt.py
#Edmund Goodman - Creative Commons Attribution-NonCommercial-ShareAlike 2.5
#Langton's Ant
import os, math

class ant:
    def __init__(self, direction, position, board):
        #Initialise the member variables
        self.direction = direction
        self.position = position
        self.board = board

    def turn(self, boardColour):
        #Change the ant's direction based on the current square colour
        if boardColour == 1:
            self.direction -= 1
        else:
            self.direction += 1
        self.direction %= 4

    def flip(self, boardColour):
        #Flip the colour of the current square
        self.board[self.position[1]][self.position[0]] = (boardColour+1)%2

    def move(self):
        #Move the ant 1 square in the current direction
        if self.direction == 0:
            self.position[1] += 1
        elif self.direction == 1:
            self.position[0] += 1
        elif self.direction == 2:
            self.position[1] -= 1
        else:
            self.position[0] -= 1

    def displayBoard(self):
        #Display the board
        for i,y in enumerate(self.board):
            for j,x in enumerate(y):
                if i == self.position[1] and j == self.position[0]:
                    print(list("^>v<")[self.direction], end="")
                elif x:
                    print("\u2588", end="")
                else:
                    print(" ", end="")
            print()

#Initialise variables
iterationCount, width, height = 12500, 150, 150

#Initialise the ant
ant1 = ant(0,[int(width/2),int(height/2)], [[0 for _ in range(height)] for _ in range(width)])

for i in range(iterationCount):
    #Take one step in the automaton
    boardColour = ant1.board[ant1.position[1]][ant1.position[0]]
    ant1.turn(boardColour)
    ant1.flip(boardColour)
    ant1.move()

#Print the board
ant1.displayBoard()

File: langtonsAnt.cc
//Edmund Goodman - Creative Commons Attribution-NonCommercial-ShareAlike 2.5
//Langton's Ant
#include<iostream>
extern const int width = 150;
extern const int height = 150;

class Ant {
    //Initialise the private variables
    int direction;
    int position[2];
    int grid[height][width];

  public:
    void generateGrid() {
      //Set every square in the grid to 0
      for (int y=0; y<height; y++) {
        for (int x=0; x<width; x++) {
          grid[y][x] = 0;
        }
      }
    }

    void setValues(int d, int p[2]) {
      //Set the private variables values
      direction = d;
      position[0] = p[0];
      position[1] = p[1];
    }

    void printGrid() {
      //Print the grid array
      for (int y=0; y<height; y++) {
        for (int x=0; x<width; x++) {
          if (x == position[0] and y == position[1]) {
            char directionIcons[5] = "^>v<";
            std::cout << directionIcons[direction];
          } else if (grid[y][x] == 0) {
            std::cout << " ";
          } else {
            std::cout << "\u2588";
          }
        }
        std::cout << std::endl;
      }
    }

    int getBoardColour() {
      //Return the board colour
      return grid[position[1]][position[0]];
    }

    void turnAnt(int boardColour) {
      //Change the ant's direction based on the current square colour
      if (boardColour == 1) {
        direction -= 1;
      } else {
        direction += 1;
      }
      if (direction<0) {
        direction = 3;
      } else if (direction>3) {
        direction = 0;
      }
    }

    void flipAntSquare(int boardColour) {
      //Flip the colour of the current square
      grid[position[1]][position[0]] = (boardColour+1)%2;
    }

    void moveAnt() {
      //Move the ant 1 square in the current direction
      if (direction == 0) {
        position[1] += 1;
      } else if (direction == 1) {
        position[0] += 1;
      } else if (direction == 2) {
        position[1] -= 1;
      } else {
        position[0] -= 1;
      }
    }
};

int main() {
  //Initialise variables
  int iterations = 12500;
  int startDirection = 0;
  int startPosition[2] = {width/2, height/2};

  //Initialise the ant
  Ant ant1;
  ant1.setValues(startDirection, startPosition);
  ant1.generateGrid();

  for (int i=0; i<iterations; i++) {
    //Take one step in the automaton
    int boardColour = ant1.getBoardColour();
    ant1.turnAnt(boardColour);
    ant1.flipAntSquare(boardColour);
    ant1.moveAnt();

  }

  //Print the board
  ant1.printGrid();

  return 0;
}
